servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultSingleAddressHttpClientBuilder.java

```diff
@@ -278,19 +281,52 @@ public StreamingHttpClient buildStreaming() {
             final HttpExecutionStrategy executionStrategy = ctx.executionContext.executionStrategy();
             // closed by the LoadBalancer
             final ConnectionFactory<R, LoadBalancedStreamingHttpConnection> connectionFactory;
+            final StreamingHttpRequestResponseFactory reqRespFactory;
             if (roConfig.isH2PriorKnowledge()) {
+                H2ProtocolConfig h2Config = roConfig.h2Config();
+                assert h2Config != null;
                 connectionFactory = new H2LBHttpConnectionFactory<>(roConfig, ctx.executionContext,
-                        ctx.builder.connectionFilterFactory, reqRespFactory,
+                        ctx.builder.connectionFilterFactory,
+                        (reqRespFactory =
+                                new DefaultStreamingHttpRequestResponseFactory(ctx.executionContext.bufferAllocator(),
+                                        h2Config.headersFactory(), HTTP_2_0)),
                         ctx.builder.influencerChainBuilder.buildForConnectionFactory(executionStrategy),
                         connectionFactoryFilter, ctx.builder.loadBalancerFactory::toLoadBalancedConnection);
             } else if (roConfig.tcpConfig().isAlpnConfigured()) {
+                H1ProtocolConfig h1Config = roConfig.h1Config();
+                H2ProtocolConfig h2Config = roConfig.h2Config();
                 connectionFactory = new AlpnLBHttpConnectionFactory<>(roConfig, ctx.executionContext,
-                        ctx.builder.connectionFilterFactory, reqRespFactory,
+                        ctx.builder.connectionFilterFactory,
+                        new AlpnReqRespFactoryFunc(ctx.executionContext.bufferAllocator(),
+                                h1Config == null ? null : h1Config.headersFactory(),
+                                h2Config == null ? null : h2Config.headersFactory()),
                         ctx.builder.influencerChainBuilder.buildForConnectionFactory(executionStrategy),
                         connectionFactoryFilter, ctx.builder.loadBalancerFactory::toLoadBalancedConnection);
+                // The client can't know which protocol will be negotiated, so just default to the preferred protocol
+                // and pay additional conversion costs if we don't get our preference.
+                String preferredAlpnProtocol = roConfig.tcpConfig().preferredAlpnProtocol();
+                if (H2_ALPN_ID.equals(preferredAlpnProtocol)) {
```

Comment:
Can you please add a test when a client has h2 as `preferredAlpnProtocol` but the server can negotiate only h1? Worry that h2 headers may not fly with h1 encoder.

Reply:
Actually, it should be already covered by `AlpnClientAndServerTest`

Reply:
discussed offline, `AlpnClientAndServerTest` uses `reserveConnection` so your original request is valid and I'll enhance this test to issue requests at the client level too (where we don't know what protocol will be negotiated).